### ISSUE TITLE
"Even Better TOML" does not recognize a standard Cargo.toml file

### ISSUE THREAD

**Issue by kkonevets** [tamasfe/taplo#387]
Type: <b>Bug</b>

Cargo.toml file lines are all red. It says there is an error somewhere but it compiles fine with cargo. Most weird error is this: "Additional properties are not allowed ('package', 'dependencies', 'features' were unexpected)"

The error:

"Categories" is a required propertyEven Better TOML
"License" is a required propertyEven Better TOML
"Builds" is a required propertyEven Better TOML
{"package":{"name":"nolik-metadata","version":"0.1.0","edition":"2021"},"dependencies":{"codec":{"package":"parity-scale-codec","version":"3.0.0","default-features":false,"features":["derive"]},"scale-info":{"version":"2.1.1","default-features":false,"features":["derive"]},"crypto_box":{"version":"0.8","optional":true},"nolik-cypher":{"path":"../cypher","optional":true},"blake2":{"version":"0.10","optional":true,"default-features":false},"getrandom":{"version":"0.2","default-features":false,"features":["custom"]},"serde":{"version":"1.0","features":["derive"],"optional":true},"serde_json":{"version":"1.0","optional":true},"base64":{"version":"0.21","optional":true},"wee_alloc":{"version":"0.4.5","optional":true}},"features":{"default":["std"],"std":["codec/std","scale-info/std","nolik-cypher","crypto_box/std","blake2","serde","serde_json","base64"],"ffi":[],"custom":["ffi","wee_alloc"]}} is not valid under any of the given schemasEven Better TOML
Additional properties are not allowed ('package', 'dependencies', 'features' were unexpected)

Extension version: 0.19.0
VS Code version: Code 1.76.1 (5e805b79fcb6ba4c2d23712967df89a089da575b, 2023-03-08T16:31:31.010Z)
OS version: Linux x64 5.4.0-144-generic
Modes:
Sandboxed: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|AMD Ryzen 9 5950X 16-Core Processor (32 x 2199)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: disabled_software<br>video_encode: disabled_software<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: disabled_off|
|Load (avg)|1, 4, 6|
|Memory (System)|31.34GB (21.24GB free)|
|Process Argv|--unity-launch --crash-reporter-id 7ca70938-323c-4f11-8635-6dc1c8549541|
|Screen Reader|no|
|VM|0%|
|DESKTOP_SESSION|mate|
|XDG_CURRENT_DESKTOP|MATE|
|XDG_SESSION_DESKTOP|mate|
|XDG_SESSION_TYPE|x11|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
vslsvsres303:30308271
pythonvspyl392:30443607
vserr242cf:30382550
pythontb:30283811
vsjup518:30340749
pythonptprofiler:30281270
vshan820:30294714
vstes263:30335439
vscoreces:30445986
pythondataviewer:30285071
vscod805cf:30301675
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
cmake_vspar411:30581797
vsaa593:30376534
pythonvs932:30410667
cppdebug:30492333
vsclangdf:30486550
c4g48928:30535728
dsvsc012cf:30540253
pynewext54:30681850
azure-dev_surveyone:30548225
pyindex848:30662994
nodejswelcome1cf:30587006
282f8724:30602487
pyind779:30671433
f6dab269:30613381
pythonsymbol12:30671437
showlangstatbar:30672706
funwalk2cf:30682975
pythonms35cf:30686773

```

</details>

![image](https://user-images.githubusercontent.com/16817116/225210078-484e1a6a-e770-4d46-ba66-047041b8df30.png)

**Comment by montchr** [comment]
It seems this isn't scoped just to Cargo.toml – I'm seeing it in a cog.toml file (configuration for [Cocogitto](https://docs.cocogitto.io/config/#general)).

**Comment by kriswuollett** [comment]
I'm getting the same errors above for a Rust `Cargo.toml` in a directory also containing the string `metadata`.

**Comment by kriswuollett** [comment]
This library uses a schema catalog, default here:

https://github.com/tamasfe/taplo/blob/55c393ec1fb84eed54933d512a4342ae54056d5a/crates/taplo-common/src/schema/associations.rs#L20

I'm wondering if file processing doesn't limit itself to checking the file type in the catalog and just validating the type as `F-Droid Data metadata` for me since this is its definition in the [schema catalog](https://github.com/SchemaStore/schemastore/blob/45db3e2f61b631b6ff47b9e6d724e4caccb1c53f/src/api/json/catalog.json#L1615) and my `Cargo.toml` is a directory named `metadata`:

```json
    {
      "name": "F-Droid Data metadata",
      "description": "Schema for F-Droid Data app metadata files",
      "fileMatch": ["**/metadata/*.yml"],
      "url": "https://gitlab.com/fdroid/fdroiddata/-/raw/master/schemas/metadata.json"
    }
```

which would be why it is complaining about [fields like `Categories`](https://gitlab.com/fdroid/fdroiddata/-/blob/a83f50c35271c3ccd7816ed6f2e7fa6b2bb79c0e/schemas/metadata.json#L12-L39):

```json
    "Categories": {
      "description": "Any number of categories for the application to be placed in.",
      "type": "array",
      "items": {
        "type": "string",
        "enum": [
          "Connectivity",
          "Development",
          "Games",
          "Graphics",
          "Internet",
          "Money",
          "Multimedia",
          "Navigation",
          "Phone & SMS",
          "Reading",
          "Science & Education",
          "Security",
          "Sports & Health",
          "System",
          "Theming",
          "Time",
          "Writing"
        ]
      },
      "uniqueItems": true,
      "minItems": 1
    },
```

So perhaps that is an earlier match than finding a [standalone `Cargo.toml` file](https://github.com/SchemaStore/schemastore/blob/45db3e2f61b631b6ff47b9e6d724e4caccb1c53f/src/api/json/catalog.json#L761-765):

```json
    {
      "name": "Cargo Manifest",
      "description": "Manifest for Cargo, the Rust package manager and build tool",
      "fileMatch": ["Cargo.toml"],
      "url": "https://json.schemastore.org/cargo.json"
    },
```

**Comment by ia0** [comment]
Did you try taplo 0.8.1? It contains #426 which should fix this issue. If this is VS Code specific, I think the version is 0.19.1. But then you may be hitting #434.

**Comment by xmlking** [comment]
this bug is blocking us to use https://github.com/cocogitto/cocogitto

**Comment by panekj** [comment]
> this bug is blocking us to use [cocogitto/cocogitto](https://github.com/cocogitto/cocogitto)

You can temporarily disable the extension

**Comment by kriswuollett** [comment]
The workaround/solution seems to be adding a [Taplo configuration file](https://github.com/tamasfe/taplo/blob/86fb2a2eedda3a0bfaef4940d4897b375c22d9e4/site/site/configuration/file.md), e.g., `.taplo.toml` in your workspace root. Wish that was pointed out earlier. 

```toml
[[rule]]
include = ["**/Cargo.toml"]

[rule.schema]
url = "https://json.schemastore.org/cargo.json"
```

I believe it works because a configuration rule has a higher priority than the catalog:

https://github.com/tamasfe/taplo/blob/42bce0c6a753a394ca40263660a889165c2762e5/crates/taplo-common/src/schema/associations.rs#L22-L31

I don't think this will ever get fixed without manual configuration because in essence it means `taplo` would have to start specifying arbitrary or an heuristic-based automatic assignment for within a schema source itself.

**Comment by jjyyxx** [comment]
After a long time of debugging, I never considered that the cause of the bug could be a folder named metadata.

**Comment by l1x** [comment]
How can I disable schema validation entirely? I am not sure how to configure which files gets which schema but it marks most of my toml files as invalid.

```
False schema does not allow "random value" Even Better TOML
```

**Comment by ia0** [comment]
Good question. I tried to find a solution in https://github.com/tamasfe/taplo/issues/457#issuecomment-1677931199 but failed. I'm not sure if it's a bug that the LSP doesn't take the `schema.enabled` setting into account. If it is, then fixing that would be a simple solution. You would then just need to add a `taplo.toml` file with:

```toml
[schema]
enabled = false
```

**Comment by kriswuollett** [comment]
As noted above in https://github.com/tamasfe/taplo/issues/387#issuecomment-1789432425, the source code shows that you can patch the detection order by supplying a config file. This is a rare issue where I'd be okay if the project closed the issue by adding to the docs. With vscode this is the entire contents of my `.taplo.toml` file within which I only needed to patch detection for Rust files which was easier than editing any config in vscode settings file(s):

```toml
[formatting]
reorder_keys = false

[[rule]]
include = ["**/Cargo.toml"]
keys = ["dependencies"]

[rule.formatting]
reorder_keys = true

[rule.schema]
url = "https://json.schemastore.org/cargo.json"
```

If you wish to keep scheme validation on, I'd assume you could just add rules as needed to push your file types to the front of the order.

**Comment by ia0** [comment]
Your comment and the original issue is different than what @l1x was asking (disabling schema validation). The original issue has actually been fixed by #426, so I'm going to close this issue. If not, please reopen with reproduction steps. Note that you should try building taplo from head, because the fix is probably not released for VS Code users (see #473).

@l1x could you open a different issue for disabling schema validation if this is still a concern for you?
